logstash-core/lib/logstash/java_pipeline.rb

```diff
@@ -376,21 +376,23 @@ def inputworker(plugin)
   # this method is intended to be called from outside the pipeline thread
   # @param before_stop [Proc] code block called before performing stop operation on input plugins
   def shutdown(&before_stop)
+    return if finished_execution?
     # shutdown can only start once the pipeline has completed its startup.
     # avoid potential race condition between the startup sequence and this
     # shutdown method which can be called from another thread at any time
     sleep(0.1) while !ready?
 
     # TODO: should we also check against calling shutdown multiple times concurrently?
 
-    before_stop.call if block_given?
+    background_thread = Thread.new { before_stop.call } if block_given?
```

Comment:
The proc that is given is called `before_stop`, but this PR changes the semantics to run the thread _in parallel_ with the stopping of the pipeline.

Consider renaming the variable to better match intent.

Reply:
I've been thinking and we should likely even remove this flexibility to pass in a block
The only time this is used is for shutdownwatcher, so we should just run it here.

Reply:
after review, this PR does not really change the execution of the shutdown watcher itself, as it always ran in a separate thread. we only skip the class level `start` method that sets up the thread for us and we do it here.

Reply:
Ah. Does this orphan `ShutdownWatcher::start` then, and if so, can we remove it?

***

> I've been thinking and we should likely even remove this flexibility to pass in a block
> The only time this is used is for shutdownwatcher, so we should ready?just run it here.

Agreed. In specs, it appears we send `Pipeline#shutdown` without a block, so we don't want to spawn a watcher in specs we would need to allow that to be configurable. Alternatively, maybe the shutdown watcher should be an attribute of the pipeline defined at initialization.

~~~ ruby
  def shutdown(shutdown_watcher_impl=nil)
    return if finished_execution?
    sleep(0.1) until ready?

    shutdown_watcher_thread = shutdown_watcher_impl && Thread.new { shutdown_watcher_impl.new(this).start }
    # ...

  ensure
    # ...
    shutdown_watcher_thread.join unless shutdown_watcher_thread.nil?
  end
~~~

This would change usage to be this with production-type code:

~~~
  pipeline.shutdown(ShutdownWatcher)
~~~

The following spec-type code would still work unmodified:

~~~
  pipeline.shutdown
~~~

***

I'm sure there are Historical Reasons, but it is odd to me that 100% of our production usage of `Pipeline#shutdown` (AFAICT?) is immediately followed by sending `Thread#join` to the result of `Pipeline#thread` (but of course, this is not done with spec code -- why would we not want to join the pipeline thread in specs?)

Reply:
@yaauie  I've removed the orphaned class method, and focused the tests on the `pipeline.start` / `pipeline.shutdown` instead of `pipeline.run`. This removed the thread juggling you mentioned in many tests, leading to a lot of deleted code.

Let me know what you think :)

Reply:
+1 to all this